File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    n, m = len(maps), len(maps[0])
    
    visited = set([(0, 0)])  # 올바르게 초기화
    queue = deque([(0, 0, 1)])  # 시작 위치와 초기 거리
    
    while queue:
        x, y, distance = queue.popleft()
        
        # 목표 위치에 도달하면 최단 거리 반환
        if x == n - 1 and y == m - 1:
            return distance
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            # 맵 범위 확인 및 방문 여부 확인
            if 0 <= nx < n and 0 <= ny < m:
                if (nx, ny) not in visited:
                    if maps[nx][ny] == 1:  # 이동 가능한 위치 확인
                        visited.add((nx, ny))
                        queue.append((nx, ny, distance + 1))  # 거리 증가

    return -1  # 도달할 수 없을 경우 -1 반환

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;

#define X first
#define Y second

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    int dx[4] = {1,0,-1,0};
    int dy[4] = {0,1,0,-1};
    int n = maps.size();
    int m = maps[0].size();
    vector<vector<int>> dist(n,vector<int>(m, -1));
    queue<pair<int,int>> Q;
    
    Q.push({0,0});
    dist[0][0] = 1;
    
    while(!Q.empty())
    {
        pair<int,int> cur = Q.front();
        Q.pop();
        for(int dir = 0; dir < 4; dir++)
        {
            int nx = cur.X + dx[dir];
            int ny = cur.Y + dy[dir];
            
            if(nx < 0 || nx >= n || ny < 0 || ny >= m)
                continue;
            if(dist[nx][ny] > 0 || maps[nx][ny] == 0)
                continue;
           
            dist[nx][ny] = dist[cur.X][cur.Y] + 1;                      
            Q.push({nx,ny});
        }
    }
    
    answer = dist[n-1][m-1];
    
    return answer;
}